File: src/main/java/com/hash_tag/GreenCup/users/UserService.java
package com.hash_tag.GreenCup.users;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;

    // Get all users
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    // Get a user by ID
    public Optional<User> getUserById(Long userId) {
        return userRepository.findById(userId);
    }

    // Create a new user
    public User createUser(User user) {
        return userRepository.save(user);
    }

    // Update an existing user
    public User updateUser(Long userId, User updatedUser) {
        return userRepository.findById(userId)
                .map(user -> {
                    user.setUsername(updatedUser.getUsername());
                    user.setPassword(updatedUser.getPassword());
                    user.setFirstName(updatedUser.getFirstName());
                    user.setLastName(updatedUser.getLastName());
                    user.setEmail(updatedUser.getEmail());
                    user.setPhone(updatedUser.getPhone());
                    user.setAddress(updatedUser.getAddress());
                    user.setCity(updatedUser.getCity());
                    user.setDateOfBirth(updatedUser.getDateOfBirth());
                    user.setGender(updatedUser.getGender());
                    user.setRole(updatedUser.getRole());
                    return userRepository.save(user);
                })
                .orElseThrow(() -> new RuntimeException("User not found with ID: " + userId));
    }

    // Delete a user by ID
    public void deleteUser(Long userId) {
        userRepository.deleteById(userId);
    }

    public Optional<User> findUserByUsername(String username) {
        return userRepository.findByUsername(username);
    }

    public Optional<List<User>> findUserByRole(String role) {
        return userRepository.findByRole(role);
    }
}

File: src/main/green-frontend/src/services/UserService.js
import axios from 'axios';

class UserService {
    constructor() {
        this.apiBaseUrl = "http://localhost:8080/user";  // Set this to your backend's base URL
    }

    // Get all users
    getAllUsers() {
        return axios.get(`${this.apiBaseUrl}/get-all-users`);
    }

    // Get a user by ID
    getUserById(userId) {
        return axios.get(`${this.apiBaseUrl}/${userId}`);
    }

    getUserByRole(role) {
        return axios.get(`${this.apiBaseUrl}/role/${role}`)
    }

    // Create a new user
    createUser(user) {
        return axios.post(`${this.apiBaseUrl}/create`, user);
    }

    // Update an existing user
    updateUser(userId, user) {
        return axios.put(`${this.apiBaseUrl}/update/${userId}`, user);
    }

    // Delete a user by ID
    deleteUser(userId) {
        return axios.delete(`${this.apiBaseUrl}/delete/${userId}`);
    }
}

export default new UserService();
